File: examples/4_time_and_state/deadlock/deadlock.py
import threading
import time

# Recursos compartidos
resource_a = threading.Lock()
resource_b = threading.Lock()

def thread_1():
    print("Thread 1: intentando adquirir Resource A...")
    resource_a.acquire()
    print("Thread 1: Resource A adquirido.")
    time.sleep(1)

    print("Thread 1: intentando adquirir Resource B...")
    resource_b.acquire()
    print("Thread 1: Resource B adquirido.")

    # Liberar recursos
    resource_b.release()
    resource_a.release()
    print("Thread 1: recursos liberados.")

def thread_2():
    print("Thread 2: intentando adquirir Resource B...")
    resource_b.acquire()
    print("Thread 2: Resource B adquirido.")
    time.sleep(1)

    print("Thread 2: intentando adquirir Resource A...")
    resource_a.acquire()
    print("Thread 2: Resource A adquirido.")

    # Liberar recursos
    resource_a.release()
    resource_b.release()
    print("Thread 2: recursos liberados.")
if __name__ == "__main__":
    # Creación de hilos
    t1 = threading.Thread(target=thread_1)
    t2 = threading.Thread(target=thread_2)

    # Iniciar hilos
    t1.start()
    t2.start()

    # Esperar a que los hilos terminen
    t1.join()
    t2.join()

    print("Programa finalizado.")

File: examples/4_time_and_state/deadlock/deadlock.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

// Recursos compartidos
pthread_mutex_t resource_a;
pthread_mutex_t resource_b;

void *thread_1(void *arg) {
    printf("Thread 1: intentando adquirir Resource A...\n");
    pthread_mutex_lock(&resource_a);
    printf("Thread 1: Resource A adquirido.\n");
    sleep(1);

    printf("Thread 1: intentando adquirir Resource B...\n");
    pthread_mutex_lock(&resource_b);
    printf("Thread 1: Resource B adquirido.\n");

    // Liberar recursos
    pthread_mutex_unlock(&resource_b);
    pthread_mutex_unlock(&resource_a);
    printf("Thread 1: recursos liberados.\n");

    return NULL;
}

void *thread_2(void *arg) {
    printf("Thread 2: intentando adquirir Resource B...\n");
    pthread_mutex_lock(&resource_b);
    printf("Thread 2: Resource B adquirido.\n");
    sleep(1);

    printf("Thread 2: intentando adquirir Resource A...\n");
    pthread_mutex_lock(&resource_a);
    printf("Thread 2: Resource A adquirido.\n");

    // Liberar recursos
    pthread_mutex_unlock(&resource_a);
    pthread_mutex_unlock(&resource_b);
    printf("Thread 2: recursos liberados.\n");

    return NULL;
}

int main() {
    pthread_t t1, t2;

    // Inicializar los mutex
    pthread_mutex_init(&resource_a, NULL);
    pthread_mutex_init(&resource_b, NULL);

    // Crear hilos
    pthread_create(&t1, NULL, thread_1, NULL);
    pthread_create(&t2, NULL, thread_2, NULL);

    // Esperar a que los hilos terminen
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    // Destruir los mutex
    pthread_mutex_destroy(&resource_a);
    pthread_mutex_destroy(&resource_b);

    printf("Programa finalizado.\n");

    return 0;
}
